src/core/common/message.hpp

```diff
@@ -217,6 +217,7 @@ class Buffer : public otMessageBuffer, public LinkedListEntry<Buffer>
         bool    mDoNotEvict : 1;       // Whether this message may be evicted.
         bool    mMulticastLoop : 1;    // Whether this multicast message may be looped back.
         bool    mResolvingAddress : 1; // Whether the message is pending an address query resolution.
+        uint8_t mOrigin : 2;           // The origin of the message.
```

Comment:
Thanks for adding this. I know the PR is still in draft mode, but I did a quick glance through and it looks good 😃 👍 .
One quick point: We should consider whether we want to copy `mOrigin` in `Message::Clone()` methods when we create a clone? I think probably we should.

Reply:
I agree with copying `mOrigin` as part of `Clone()`.

Reply:
Thanks for pointing it out! I didn't realized I need to manually copy the metadata fields.